File: packages/java/src/main/java/com/yourdudeken/mpesa/auth/AuthService.java
package com.yourdudeken.mpesa.auth;

import com.yourdudeken.mpesa.config.MpesaConfig;
import java.util.HashMap;
import java.util.Map;

public class AuthService {
    private MpesaConfig config;
    private String baseUrl;
    private String accessToken;
    private long tokenExpiry;

    public AuthService(MpesaConfig config) {
        this.config = config;
        this.baseUrl = config.getEnvironment().equals("sandbox")
            ? "https://sandbox.safaricom.co.ke"
            : "https://api.safaricom.co.ke";
    }

    public String getAccessToken(String shortCodeType) {
        if (accessToken != null && System.currentTimeMillis() < tokenExpiry) {
            return accessToken;
        }
        return generateAccessToken(shortCodeType);
    }

    private String generateAccessToken(String shortCodeType) {
        String consumerKey = shortCodeType.equals("B2C") || shortCodeType.equals("B2B")
            ? config.getB2cConsumerKey()
            : config.getMpesaConsumerKey();
        String consumerSecret = shortCodeType.equals("B2C") || shortCodeType.equals("B2B")
            ? config.getB2cConsumerSecret()
            : config.getMpesaConsumerSecret();

        String auth = consumerKey + ":" + consumerSecret;
        String encodedAuth = java.util.Base64.getEncoder().encodeToString(auth.getBytes());

        try {
            java.net.URL url = new java.net.URL(baseUrl + "/oauth/v1/generate?grant_type=client_credentials");
            java.net.HttpURLConnection conn = (java.net.HttpURLConnection) url.openConnection();
            conn.setRequestMethod("GET");
            conn.setRequestProperty("Authorization", "Basic " + encodedAuth);

            java.io.BufferedReader reader = new java.io.BufferedReader(
                new java.io.InputStreamReader(conn.getInputStream())
            );
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line);
            }
            reader.close();

            com.google.gson.JsonObject json = new com.google.gson.JsonParser().parse(response.toString()).getAsJsonObject();
            accessToken = json.get("access_token").getAsString();
            tokenExpiry = System.currentTimeMillis() + (json.get("expires_in").getAsLong() - 60) * 1000;

            return accessToken;
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate access token: " + e.getMessage());
        }
    }

    public void clearToken() {
        accessToken = null;
        tokenExpiry = 0;
    }
}

File: packages/dotnet/src/Yourdudeken.Mpesa/Auth/AuthService.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

namespace Yourdudeken.Mpesa.Auth
{
    public class AuthService
    {
        private readonly Config.MpesaConfig _config;
        private string _baseUrl;
        private string _accessToken;
        private long _tokenExpiry;

        public AuthService(Config.MpesaConfig config)
        {
            _config = config;
            _baseUrl = config.Environment == "sandbox"
                ? "https://sandbox.safaricom.co.ke"
                : "https://api.safaricom.co.ke";
        }

        public async Task<string> GetAccessToken(string shortCodeType = "C2B")
        {
            if (!string.IsNullOrEmpty(_accessToken) && DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() < _tokenExpiry)
            {
                return _accessToken;
            }

            return await GenerateAccessToken(shortCodeType);
        }

        private async Task<string> GenerateAccessToken(string shortCodeType)
        {
            var consumerKey = shortCodeType == "B2C" || shortCodeType == "B2B"
                ? _config.B2cConsumerKey
                : _config.MpesaConsumerKey;
            var consumerSecret = shortCodeType == "B2C" || shortCodeType == "B2B"
                ? _config.B2cConsumerSecret
                : _config.MpesaConsumerSecret;

            var auth = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{consumerKey}:{consumerSecret}"));

            using var client = new HttpClient();
            client.DefaultRequestHeaders.Add("Authorization", $"Basic {auth}");

            var response = await client.GetAsync($"{_baseUrl}/oauth/v1/generate?grant_type=client_credentials");
            var content = await response.Content.ReadAsStringAsync();

            var json = System.Text.Json.JsonDocument.Parse(content);
            _accessToken = json.RootElement.GetProperty("access_token").GetString();
            var expiresIn = json.RootElement.GetProperty("expires_in").GetInt64();
            _tokenExpiry = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + (expiresIn - 60) * 1000;

            return _accessToken;
        }

        public void ClearToken()
        {
            _accessToken = null;
            _tokenExpiry = 0;
        }
    }
}
